YKI(Frontend): OPHYKIKEH-307 Show user's active registrations in header

lket opened the pull request:
(Liittyy: https://github.com/Opetushallitus/yki/pull/229)

Tämä on vielä vähän raakile. Tarkoitus oli vielä parantaa rajatapauksien hallintaa (erityisesti tilan päivittäminen palvelimelta sopivissa kohdissa), mutta aika jäi vähän kesken.

Avaan tämän nyt tällaisena draftina, jos tätä haluaa tutkiskella lomani aikana.

Erityisesti kuulisin saisiko tuon vähän turhan monimutkainen useEffects-kohdan tehtyä siistimmin. Aavistelisin että tähän olisi joku hyvä ideomaattinen saga-tapa, mutta en vielä ainakaan keksinyt sitä.

TODO:
- [x] Testit
- [x] Joko ajoittainen päivitys, tai sitten haetaan palvelimelta suoraan kauanko rekisteröinti on voimassa
- [x] Muut tilanteet jossa tämä nykyinen tieto pitää nollata tai hakea tieto uudestaan.

pkoivisto commented on frontend/packages/yki/src/components/layouts/SessionStateHeader.tsx:
```diff
@@ -39,12 +42,50 @@ const generateLogoutURL = () => {
+  const notOnRegistrationPage =
+    matchPath(AppRoutes.ExamSessionRegistration, location.pathname) === null;
+
+  useEffect(() => {
+    const needsUserOpenRegistrations =
+      !hasNullIdentity &&
+      notOnRegistrationPage &&
+      userOpenRegistrations.status === APIResponseStatus.NotStarted;
+
+    if (needsUserOpenRegistrations) {
+      dispatch(loadUserOpenRegistrations());
+    }
+  }, [hasNullIdentity, notOnRegistrationPage, userOpenRegistrations, dispatch]);
```
Oliko se tämä kohta jonka idiomaattisuus epäilytti? Näyttää ainakin ihan ok:lta ja suoraviivaiselta mun silmään.

pkoivisto commented on frontend/packages/yki/src/components/layouts/SessionStateHeader.tsx:
```diff
@@ -39,12 +42,50 @@ const generateLogoutURL = () => {
   return (
     <div className="session-header columns gapped">
+      {notOnRegistrationPage &&
+        userOpenRegistrations.openRegistrations &&
+        userOpenRegistrations.openRegistrations['open-registrations'] &&
+        userOpenRegistrations.openRegistrations['open-registrations'][0] && (
```
Syntaksi voisi helpottua jos tuon UserOpenRegistrationResponsen deserialisoisi hieman eri muotoon. Esim. tuo `userOpenRegistrations.openRegistrations` voisi jo suoraan sisältää sen kaivatun arrayn.

pkoivisto commented on frontend/packages/yki/public/i18n/fi-FI/common.json:
```diff
@@ -70,6 +70,8 @@
           "langSelectorAriaLabel": "Kieli / Språk / Language"
         },
         "sessionState": {
+          "activeRegistrations": "Sinulla on 1 käynnissä oleva ilmoittautuminen.",
```
Pitäisi hanskata myös ne tapaukset joissa on useita ilmoittautumisia käynnissä. Vaihtoehtoisesti voitaisiin varmaan bäkkärissä rajoittaa ettei oikeasti voi olla kuin yksin avoin ilmoittautuminen kerrallaan käynnissä.

lket replied:
Tuntui jotenkin aika kummalliselta että siinä ilmoitettaisiin jotenkin "Sinulla on 342 käynnissä olevaa ilmoittautumista [Linkki yhteen niistä]". Mutta toisaalta, voihan se niinkin olla kunnes tuo rajoitus mietitään jotenkin järkevästi.

lket committed:
YKI(Frontend): Reset open registrations when one expires

lket committed:
YKI(Frontend): Show user's active registrations in header

lket committed:
YKI(Frontend): Update Header snapshot test

pkoivisto commented on frontend/packages/yki/src/redux/sagas/userOpenRegistrations.ts:
```diff
@@ -0,0 +1,56 @@
+    const refreshDelay = Math.min(10 * 60 * 1000, expires_in);
+
+    yield delay(refreshDelay);
+    yield put(resetUserOpenRegistrations());
```
Tämä vissiin tyhjentää avoimet ilmoittautumiset mahdollisesti ennenaikaisesti, mistä ehkä aiheutuu käliin pientä flickeriä? Voisi ehkä harkita varianttia joka vain laukaisee päivitysprosessin uudestaan ilman että avoimia rekisteröintejä tyhjennetään ennen palvelimen vastausta.

pkoivisto commented on frontend/packages/yki/src/redux/sagas/userOpenRegistrations.ts:
```diff
@@ -0,0 +1,56 @@
+
+        return expires_at < min ? expires_at : min;
+      },
+      new Date(8640000000000000)
```
Reducen alkuarvon voisi nimetä jotenkin kuvaavasti. Mahdollisesti alkuarvo voisi myös staattisen arvon sijasta olla X minuuttia nykyhetkestä.

lket committed:
YKI(Frontend): Improve user open registrations refresh handling

- Only queue API call, do not clear the current state
- open-registrations -> open_registrations
- Clean delay handling by using dayjs

pkoivisto commented on frontend/packages/yki/src/redux/sagas/userOpenRegistrations.ts:
```diff
@@ -0,0 +1,56 @@
+    // When expiry is very close or in the past, check every minute
+    const expires_in = Math.max(60 * 1000, nextToExpire.diff(dayjs()));
+
+    // Check at least every 10 minutes
+    const refreshDelay = Math.min(1 * 60 * 1000, expires_in);
```
Kommentissa mainittu `every 10 minutes`, mutta koodissa joka minuutti?

lket replied:
Hyvä huomio, meni testimuutos sisään eilen.

Sinänsähän nämä on ihan mielivaltaisia arvoja, lähinnähän tämän koko hässäkän tarkoitus on että kaikki clientit ei hakkaisi palvelinta ihan jatkuvasti, mutta en kyllä tiedä kuinka paljon väliä sillä edes lopulta on.

lket commented:
Pienenä demoefektinä tuli vielä joku ongelma tiedon päivittymisessä sivusiirtymässä - arvelen että vanha ongelma joka tuli esiin nyt, mutta hyvä varmaan korjata nyt joka tapauksessa.

lket committed:
YKI(Frontend): Reset exam session state on registration cancel

This allows correctly navigating to different registration details
page. Previously the old exam session and new registration were in conflict.

lket commented:
Tuo uusin commit mielestäni korjaa demossa tulleen ongelman. Tämän pitäisi riittää korjaukseksi, kunhan rekisteröintilomakkeelta poistutaan yhden sivulatauksen aikana vain rekisteröinnin perumisen kautta - kuten minun nähdäkseni asia on. Jos siihen joskus tulee muutos, pitää tilan käsittelyyn ehkä tehdä joku isompi muutos.

pkoivisto commented on frontend/packages/yki/src/components/layouts/SessionStateHeader.tsx:
```diff
@@ -39,12 +42,49 @@ const generateLogoutURL = () => {
   const session = useAppSelector(sessionSelector).loggedInSession;
-  if (!session || session.identity === null) return <></>;
+  const dispatch = useAppDispatch();
+  const hasNullIdentity = !session || session.identity === null;
```
Jäin pohtimaan olisiko luontevampaa kääntää tämän booleanin merkitys ja nimi; esim. siis `const hasSessionIdentity = ...` tai jopa `const isAuthenticated = ...`.

Ei sinänsä iso asia, voi jäädä näinkin.

lket committed:
YKI(Frontend): Simplify isAuthenticated check
